insolar/bus/bus.go

```diff
@@ -0,0 +1,163 @@
+//
+// Copyright 2019 Insolar Technologies GbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package bus
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/message/router/middleware"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/pkg/errors"
+)
+
+const (
+	// TopicOutgoing is topic for external calls
+	TopicOutgoing = "TopicOutgoing"
+
+	// TopicIncoming is topic for incoming calls
+	TopicIncoming = "TopicIncoming"
+
+	// MetaPulse is key for Pulse
+	MetaPulse = "pulse"
+
+	// MetaType is key for Type
+	MetaType = "type"
+
+	// MetaReceiver is key for Receiver
+	MetaReceiver = "receiver"
+
+	// MetaSender is key for Sender
+	MetaSender = "sender"
+)
+
+//go:generate minimock -i github.com/insolar/insolar/insolar/bus.Sender -o ./ -s _mock.go
+
+// Sender interface sends messages by watermill.
+type Sender interface {
+	// Send an `Message` and get a `Reply` or error from remote host.
+	Send(ctx context.Context, msg *message.Message) (<-chan *message.Message, func())
+}
+
+type lockedReply struct {
+	mutex    sync.RWMutex
+	messages chan *message.Message
+
+	once sync.Once
+	done chan struct{}
+}
+
+// Bus is component that sends messages and gives access to replies for them.
+type Bus struct {
+	pub     message.Publisher
+	timeout time.Duration
+
+	repliesMutex sync.RWMutex
+	replies      map[string]*lockedReply
+}
+
+// NewBus creates Bus instance with provided values.
+func NewBus(pub message.Publisher) *Bus {
+	return &Bus{
+		timeout: time.Minute * 10,
```

Comment:
Might be move it to config?

Reply:
Config should contain options for user to set, this is not one of such options.